crates/subspace-farmer/src/rpc_client.rs

```diff
@@ -4,11 +4,131 @@ use subspace_core_primitives::BlockNumber;
 use subspace_rpc_primitives::{
     BlockSignature, BlockSigningInfo, FarmerMetadata, SlotInfo, SolutionResponse,
 };
-use tokio::sync::mpsc::Receiver;
+use tokio::{
+    sync::mpsc::{self, Receiver},
+    time::Instant,
+};
 
 /// To become error type agnostic
 pub type Error = Box<dyn std::error::Error + Send + Sync + 'static>;
 
+pub enum SegmentPipelineEventSender {
```

Comment:
"sender" is an implementation detail. It might have been easier to use if it was called something like `SegmentPipelineEventTracer` with private `Option` instead of two public variants and methods of `SegmentPipelineEvent` on this data structure directly.